From 4e8e11af697749cf5bfdd606d34510156004691e Mon Sep 17 00:00:00 2001
From: Phil Sutter <phil@nwl.cc>
Date: Mon, 1 Jun 2026 01:45:09 +0200
Subject: [PATCH] mips: dts: Add rb4xx DTS

Copied from openwrt HEAD at c9406df91836b5c94a395b61ebe5c16e08e2c779
plus the following fixes:

* Add missing data to rb4xx nand definition

Engine type should not be necessary since rawnand_dt_init() sets
chip->ecc.engine_type only if the DTB specified it and nand_rb4xx sets
it correctly. Define it anyway for the sake of completeness.

* Set rb493g MMC voltage ranges

The old platform device in mach-rb4xx.c had an ocr_mask of
'MMC_VDD_32_33 | MMC_VDD_33_34', this looks like the correct equivalent.

* Drop fixed rb4xx-gpio base

Effectively convert the device to dynamic base allocation this way,
eliminating the warning about deprecated static allocation. All this is
not relevant anyway since rb4xx-gpio merely cares about the offset and
passes it on to rb4xx-cpld (spi) driver.

Signed-off-by: Phil Sutter <phil@nwl.cc>
---
 arch/mips/boot/dts/Makefile                   |   1 +
 arch/mips/boot/dts/mikrotik/Makefile          |   1 +
 arch/mips/boot/dts/mikrotik/ar7100.dtsi       | 242 ++++++++++++++++++
 .../ar7100_mikrotik_routerboard-4xx.dtsi      | 167 ++++++++++++
 .../ar7161_mikrotik_routerboard-493g.dts      |  53 ++++
 arch/mips/boot/dts/mikrotik/ath79.dtsi        |  95 +++++++
 6 files changed, 559 insertions(+)
 create mode 100644 arch/mips/boot/dts/mikrotik/Makefile
 create mode 100644 arch/mips/boot/dts/mikrotik/ar7100.dtsi
 create mode 100644 arch/mips/boot/dts/mikrotik/ar7100_mikrotik_routerboard-4xx.dtsi
 create mode 100644 arch/mips/boot/dts/mikrotik/ar7161_mikrotik_routerboard-493g.dts
 create mode 100644 arch/mips/boot/dts/mikrotik/ath79.dtsi

diff --git a/arch/mips/boot/dts/Makefile b/arch/mips/boot/dts/Makefile
index 36c5e4c6e6406..036f7534ebf77 100644
--- a/arch/mips/boot/dts/Makefile
+++ b/arch/mips/boot/dts/Makefile
@@ -7,6 +7,7 @@ subdir-y	+= img
 subdir-y	+= ingenic
 subdir-y	+= lantiq
 subdir-y	+= loongson
+subdir-y	+= mikrotik
 subdir-y	+= mscc
 subdir-y	+= mti
 subdir-y	+= ni
diff --git a/arch/mips/boot/dts/mikrotik/Makefile b/arch/mips/boot/dts/mikrotik/Makefile
new file mode 100644
index 0000000000000..6fec89e7c9a27
--- /dev/null
+++ b/arch/mips/boot/dts/mikrotik/Makefile
@@ -0,0 +1 @@
+dtb-$(CONFIG_ATH79)	+= ar7161_mikrotik_routerboard-493g.dtb
diff --git a/arch/mips/boot/dts/mikrotik/ar7100.dtsi b/arch/mips/boot/dts/mikrotik/ar7100.dtsi
new file mode 100644
index 0000000000000..beb52a16855a3
--- /dev/null
+++ b/arch/mips/boot/dts/mikrotik/ar7100.dtsi
@@ -0,0 +1,242 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "ath79.dtsi"
+
+/ {
+	compatible = "qca,ar7100";
+
+	chosen {
+		bootargs = "console=ttyS0,115200";
+	};
+
+	cpus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		cpu@0 {
+			device_type = "cpu";
+			compatible = "mips,mips24Kc";
+			clocks = <&pll ATH79_CLK_CPU>;
+			reg = <0>;
+		};
+	};
+
+	extosc: ref {
+		compatible = "fixed-clock";
+		#clock-cells = <0>;
+		clock-output-names = "ref";
+		clock-frequency = <40000000>;
+	};
+
+	ahb {
+		apb {
+			ddr_ctrl: memory-controller@18000000 {
+				compatible = "qca,ar7100-ddr-controller";
+				reg = <0x18000000 0x100>;
+
+				#qca,ddr-wb-channel-cells = <1>;
+			};
+
+			uart: uart@18020000 {
+				compatible = "ns16550a";
+				reg = <0x18020000 0x20>;
+				interrupts = <3>;
+
+				clocks = <&pll ATH79_CLK_AHB>;
+				clock-names = "uart";
+
+				reg-io-width = <4>;
+				reg-shift = <2>;
+				no-loopback-test;
+			};
+
+			usb_phy: usb-phy@18030000 {
+				compatible = "qca,ar7100-usb-phy";
+				reg = <0x18030000 0x10>;
+
+				reset-names = "phy", "host", "usb-ohci-dll";
+				resets = <&rst 4>, <&rst 5>, <&rst 6>;
+
+				#phy-cells = <0>;
+
+				status = "disabled";
+			};
+
+			gpio: gpio@18040000 {
+				compatible = "qca,ar7100-gpio";
+				reg = <0x18040000 0x28>;
+				interrupts = <2>;
+
+				ngpios = <12>;
+
+				gpio-controller;
+				#gpio-cells = <2>;
+
+				interrupt-controller;
+				#interrupt-cells = <2>;
+			};
+
+			pll: pll-controller@18050000 {
+				compatible = "qca,ar7100-pll", "syscon";
+				reg = <0x18050000 0x20>;
+
+				clocks = <&extosc>;
+				clock-names = "ref";
+
+				#clock-cells = <1>;
+				clock-output-names = "cpu", "ddr", "ahb";
+			};
+
+			wdt: wdt@18060008 {
+				compatible = "qca,ar7130-wdt";
+				reg = <0x18060008 0x8>;
+
+				interrupts = <4>;
+
+				clocks = <&pll ATH79_CLK_AHB>;
+				clock-names = "wdt";
+			};
+
+			pci_intc: interrupt-controller@18060018 {
+				compatible = "qca,ar7100-misc-intc";
+				reg = <0x18060018 0x4>;
+				interrupt-parent = <&cpuintc>;
+				interrupts = <2>;
+				interrupt-controller;
+				#interrupt-cells = <1>;
+			};
+
+			rst: reset-controller@18060024 {
+				compatible = "qca,ar7100-reset";
+				reg = <0x18060024 0x4>;
+
+				#reset-cells = <1>;
+			};
+
+			pcie0: pcie@17010000 {
+				compatible = "qca,ar7100-pci";
+				#address-cells = <3>;
+				#size-cells = <2>;
+				bus-range = <0x0 0x0>;
+				reg = <0x17010000 0x100>;
+				reg-names = "cfg_base";
+				ranges = <0x2000000 0 0x10000000 0x10000000 0 0x07000000	/* pci memory */
+					  0x1000000 0 0x00000000 0x0000000 0 0x000001>;		/* io space */
+
+				device_type = "pci";
+
+				interrupt-parent = <&pci_intc>;
+				interrupts = <4>;
+
+				#interrupt-cells = <1>;
+
+				interrupt-map-mask = <0xf800 0 0 0>;
+				interrupt-map = <0x8800 0 0 0 &pci_intc 0
+						 0x9000 0 0 0 &pci_intc 1
+						 0x9800 0 0 0 &pci_intc 2>;
+
+				status = "disabled";
+			};
+		};
+	};
+
+	usb2: usb@1b000000 {
+		compatible = "generic-ehci";
+		reg = <0x1b000000 0x1000>;
+
+		interrupt-parent = <&cpuintc>;
+		interrupts = <3>;
+
+		phy-names = "usb";
+		phys = <&usb_phy>;
+
+		has-synopsys-hc-bug;
+
+		status = "disabled";
+
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		usb_ehci_port: port@1 {
+			reg = <1>;
+			#trigger-source-cells = <0>;
+		};
+	};
+
+	usb1: usb@1c000000 {
+		compatible = "generic-ohci";
+		reg = <0x1c000000 0x1000>;
+
+		interrupt-parent = <&miscintc>;
+		interrupts = <6>;
+
+		phy-names = "usb";
+		phys = <&usb_phy>;
+
+		status = "disabled";
+
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		usb_ohci_port: port@1 {
+			reg = <1>;
+			#trigger-source-cells = <0>;
+		};
+	};
+
+	spi: spi@1f000000 {
+		compatible = "qca,ar7100-spi";
+		reg = <0x1f000000 0x10>;
+
+		clocks = <&pll ATH79_CLK_AHB>;
+		clock-names = "ahb";
+
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		status = "disabled";
+	};
+};
+
+&cpuintc {
+	qca,ddr-wb-channel-interrupts = <2>, <3>, <4>, <5>;
+	qca,ddr-wb-channels = <&ddr_ctrl 3>, <&ddr_ctrl 2>,
+				<&ddr_ctrl 0>, <&ddr_ctrl 1>;
+};
+
+&miscintc {
+	compatible = "qca,ar7100-misc-intc";
+};
+
+&eth0 {
+	compatible = "qca,ar7100-eth", "syscon";
+	reg = <0x19000000 0x200>, <0x18070000 0x4>;
+
+	pll-data = <0x00110000 0x00001099 0x00991099>;
+	pll-reg = <0x4 0x10 17>;
+	pll-handle = <&pll>;
+	phy-mode = "rgmii";
+
+	resets = <&rst 9>;
+	reset-names = "mac";
+	qca,mac-idx = <0>;
+};
+
+&mdio1 {
+	builtin-switch;
+};
+
+&eth1 {
+	compatible = "qca,ar7100-eth", "syscon";
+	reg = <0x1a000000 0x200>, <0x18070004 0x4>;
+
+	pll-data = <0x00110000 0x00001099 0x00991099>;
+	pll-reg = <0x4 0x14 19>;
+	pll-handle = <&pll>;
+
+	phy-mode = "rgmii";
+
+	resets = <&rst 13>;
+	reset-names = "mac";
+	qca,mac-idx = <1>;
+};
diff --git a/arch/mips/boot/dts/mikrotik/ar7100_mikrotik_routerboard-4xx.dtsi b/arch/mips/boot/dts/mikrotik/ar7100_mikrotik_routerboard-4xx.dtsi
new file mode 100644
index 0000000000000..8a6b4cb682ee5
--- /dev/null
+++ b/arch/mips/boot/dts/mikrotik/ar7100_mikrotik_routerboard-4xx.dtsi
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: GPL-1.0-or-later OR MIT
+
+#include "ar7100.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	aliases {
+		led-boot = &led_user;
+		led-failsafe = &led_user;
+		led-running = &led_user;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_user: user {
+			label = "yellow:user";
+			gpios = <&gpio 4 GPIO_ACTIVE_HIGH>;
+			default-state = "on";
+		};
+
+		led1 {
+			label = "green:led1";
+			gpios = <&cpld_gpio 0 GPIO_ACTIVE_LOW>;
+		};
+
+		led2 {
+			label = "green:led2";
+			gpios = <&cpld_gpio 2 GPIO_ACTIVE_LOW>;
+		};
+
+		led3 {
+			label = "green:led3";
+			gpios = <&cpld_gpio 1 GPIO_ACTIVE_LOW>;
+		};
+
+		led4 {
+			label = "green:led4";
+			gpios = <&cpld_gpio 3 GPIO_ACTIVE_LOW>;
+		};
+
+		led5 {
+			label = "green:led5";
+			gpios = <&cpld_gpio 8 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+&gpio {
+	gpio-line-names =
+		"",     "", "", "", "LED", "RDY",  "",  "MDC",
+		"MDIO", "", "", "";
+};
+
+&pcie0 {
+	status = "okay";
+};
+
+&spi {
+	status = "okay";
+
+	compatible = "mikrotik,rb4xx-spi";
+
+	flash@0 {
+		compatible = "pm25lv512", "jedec,spi-nor";
+
+		reg = <0>;
+		spi-max-frequency = <25000000>;
+
+		partitions {
+			compatible = "mikrotik,routerboot-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "routerboot1";
+				reg = <0x0 0x0>;
+				read-only;
+			};
+
+			hard_config {
+				read-only;
+			};
+
+			bios {
+				size = <0x1000>;
+				read-only;
+			};
+
+			routerboot2 {
+				label = "routerboot2";
+				read-only;
+			};
+
+			soft_config {
+			};
+		};
+	};
+
+	cpld@1 {
+		compatible = "mikrotik,rb4xx-cpld";
+
+		reg = <1>;
+		spi-max-frequency = <25000000>;
+
+		cpld_gpio: gpio {
+			compatible = "mikrotik,rb4xx-gpio";
+
+			gpio-controller;
+			#gpio-cells = <2>;
+			gpio-line-names =
+				"LED1", "LED2", "LED3", "LED4",
+				"FAN",  "ALE",  "CLE",  "nCE",
+				"LED5";
+		};
+
+		nand {
+			compatible = "mikrotik,rb4xx-nand";
+
+			gpios = <&cpld_gpio 5 GPIO_ACTIVE_HIGH>, // ALE
+				<&cpld_gpio 6 GPIO_ACTIVE_HIGH>, // CLE
+				<&cpld_gpio 7 GPIO_ACTIVE_HIGH>, // nCE
+				<&gpio      5 GPIO_ACTIVE_HIGH>; // RDY
+
+			nand-use-soft-ecc-engine;
+			nand-ecc-algo = "hamming";
+
+			partitions {
+				compatible = "fixed-partitions";
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				partition@0 {
+					label = "booter";
+					reg = <0x0000000 0x0040000>;
+					read-only;
+				};
+
+				partition@40000 {
+					label = "kernel";
+					reg = <0x0040000 0x0800000>;
+				};
+
+				partition@840000 {
+					label = "ubi";
+					reg = <0x0840000 0x77c0000>;
+				};
+			};
+		};
+	};
+};
+
+&eth0 {
+	status = "okay";
+
+	phy-handle = <&phy0>;
+};
+
+&mdio0 {
+	status = "okay";
+
+	phy0: ethernet-phy@0 {
+		reg = <0>;
+	};
+};
diff --git a/arch/mips/boot/dts/mikrotik/ar7161_mikrotik_routerboard-493g.dts b/arch/mips/boot/dts/mikrotik/ar7161_mikrotik_routerboard-493g.dts
new file mode 100644
index 0000000000000..8e33389ce8699
--- /dev/null
+++ b/arch/mips/boot/dts/mikrotik/ar7161_mikrotik_routerboard-493g.dts
@@ -0,0 +1,53 @@
+// SPDX-License-Identifier: GPL-1.0-or-later OR MIT
+
+#include "ar7100_mikrotik_routerboard-4xx.dtsi"
+
+/ {
+	compatible = "mikrotik,routerboard-493g", "qca,ar7161";
+	model = "MikroTik RouterBOARD 493G";
+
+	aliases {
+		mdio-gpio0 = &mdio_gpio0;
+	};
+};
+
+&spi {
+	sdcard: mmc-slot@2 {
+		compatible = "mmc-spi-slot";
+
+		reg = <2>;
+		spi-max-frequency = <25000000>;
+		voltage-ranges = <3200 3400>;
+	};
+};
+
+&eth1 {
+	status = "okay";
+
+	phy-handle = <&gpio_phy0>;
+
+	mdio_gpio0: mdio-gpio {
+		compatible = "virtual,mdio-gpio";
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		gpios = <&gpio 7 GPIO_ACTIVE_HIGH>, // MDC
+			<&gpio 8 GPIO_ACTIVE_HIGH>; // MDIO
+
+		gpio_phy0: ethernet-phy@0 {
+			reg = <0>;
+		};
+	};
+};
+
+&usb_phy {
+	status = "okay";
+};
+
+&usb2 {
+	status = "okay";
+};
+
+&usb1 {
+	status = "okay";
+};
diff --git a/arch/mips/boot/dts/mikrotik/ath79.dtsi b/arch/mips/boot/dts/mikrotik/ath79.dtsi
new file mode 100644
index 0000000000000..8849d729ac7d4
--- /dev/null
+++ b/arch/mips/boot/dts/mikrotik/ath79.dtsi
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
+
+#include <dt-bindings/clock/ath79-clk.h>
+
+/ {
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	cpuintc: interrupt-controller {
+		compatible = "qca,ar7100-cpu-intc";
+
+		interrupt-controller;
+		#interrupt-cells = <1>;
+	};
+
+	ahb {
+		compatible = "simple-bus";
+		ranges;
+
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		interrupt-parent = <&cpuintc>;
+
+		apb {
+			compatible = "simple-bus";
+			ranges;
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			interrupt-parent = <&miscintc>;
+
+			miscintc: interrupt-controller@18060010 {
+				compatible = "qca,ar7240-misc-intc";
+				reg = <0x18060010 0x4>;
+
+				interrupt-parent = <&cpuintc>;
+				interrupts = <6>;
+
+				interrupt-controller;
+				#interrupt-cells = <1>;
+			};
+		};
+
+		eth0: eth@19000000 {
+			status = "disabled";
+
+			compatible = "qca,ath79-eth", "syscon";
+			reg = <0x19000000 0x200>;
+
+			interrupts = <4>;
+			phy-mode = "mii";
+			syscon-no-reset;
+
+			mdio0: mdio {
+				status = "disabled";
+
+				compatible = "qca,ath79-mdio";
+				#address-cells = <1>;
+				#size-cells = <0>;
+
+				regmap = <&eth0>;
+
+				clocks = <&pll ATH79_CLK_MDIO>;
+				clock-names = "ref";
+			};
+		};
+
+		eth1: eth@1a000000 {
+			status = "disabled";
+
+			compatible = "qca,ath79-eth", "syscon";
+			reg = <0x1a000000 0x200>;
+
+			interrupts = <5>;
+			phy-mode = "mii";
+			syscon-no-reset;
+
+			mdio1: mdio {
+				status = "disabled";
+
+				compatible = "qca,ath79-mdio";
+				#address-cells = <1>;
+				#size-cells = <0>;
+
+				regmap = <&eth1>;
+
+				clocks = <&pll ATH79_CLK_MDIO>;
+				clock-names = "ref";
+			};
+		};
+	};
+};
